server/api/Users.js

```diff
@@ -9,64 +9,76 @@ const {
   } catch (er) {
     next(er);
   }
 });
 
-// GET /api/users
-router.get('/', requireUserToken, async (req, res, next) => {
-  const { user } = req;
-  try {
-    console.log('----->GET/ api/users', user);
-    res.send(user);
-  } catch (er) {
-    next(er);
-  }
-});
-
-// GET /api/users/:id/cart
-// retrieves a cart's associated products from database
-router.get('/:id/cart', requireCartToken, async (req, res, next) => {
-  const { cart } = req;
-  console.log('requireCartToken', cart);
+// DELETE /api/users/:id deletes a user if given a valid Admin token
+router.delete('/:id', requireAdminToken, async (req, res, next) => {
   try {
-    res.send(cart);
+    const {
+      params: { id },
+    } = req;
+    const user = await User.findByPk(id);
+    await user.destroy();
+    res.sendStatus(204);
   } catch (er) {
     next(er);
   }
 });
 
-// PUT /api/users/:id
-router.put('/:id', async (req, res, next) => {
+// GET /api/users returns the current user details with a valid User token
+router.get('/', requireUserToken, async (req, res, next) => {
+  const { user } = req;
   try {
-    const user = await User.findByPk(req.params.id);
-    res.status(200).send(await user.update(req.body));
+    // console.log('----->GET/ api/users', user);
+    res.send(user);
   } catch (er) {
     next(er);
   }
 });
 
-// DELETE /api/users/:id
-router.delete('/:id', async (req, res, next) => {
+// PUT /api/users updates the current user details with a valid User token
+router.put('/', requireUserToken, async (req, res, next) => {
   try {
-    const user = await User.findByPk(req.params.id);
-    await user.destroy();
-    res.sendStatus(204);
+    let { user } = req;
+    const { email, password, firstName, lastName, phoneNumber } = req.body;
+    user.email = email || user.email;
+    user.password = password || user.password;
+    user.firstName = firstName || user.firstName;
+    user.lastName = lastName || user.lastName;
+    user.phoneNumber = phoneNumber || user.phoneNumber;
+    await user.save();
+    user = await User.findByPk(user.id, {
+      attributes: { exclude: ['password'] },
+    });
+    res.send(user);
   } catch (er) {
     next(er);
   }
 });
 
-// POST /api/users
-router.post('/', async (req, res, next) => {
+// POST /api/users creates a new user with the details provided
+router.post('/', requireCartToken, async (req, res, next) => {
   try {
-    res.status(201).send(await User.create(req.body));
+    const { cart } = req;
+    const { email, firstName, lastName, phoneNumber, password } = req.body;
+    const newUserDetails = { email, firstName, lastName, phoneNumber };
+    newUserDetails.cartId = cart.id;
+    let user = await User.create(newUserDetails);
+    await user.setCart(cart);
+    if (password) {
+      await user.update({ password });
+    }
+    user = await User.findByPk(user.id, {
+      attributes: { exclude: ['password'] },
+    });
+    res.send(user);
   } catch (er) {
     next(er);
   }
```

Comment:
`POST /api/users` requires a valid Cart token (which any visitor to our site should get when they first arrive) and creates a new User instance, with email, firstName, lastName, phoneNumber and optionally password. It also links the new user to the cart provided by the requireCartToken utility. It then sends back the new user without their password attribute.